### PR TITLE
First commit!

### PR DESCRIPTION
**Pull request by colinmorris**
Still a bit rough in places, but it's probably not going to get any cleaner from this point on, and the amount of time I've spent in this branch is kind of silly.

Main changes:
- now there's computer code

### CODE REVIEW

File: rbm_softmax.py [hunk 1]
```diff
@@ -0,0 +1,504 @@
+"""Restricted Boltzmann Machine with softmax visible units.
```
Comment (colinmorris) [inline rbm_softmax.py:1]: I'm actually going to push the corresponding sklearn file to master, so it's clear which parts here I added, and what's cannibalized (a lot).

Edit: Why isn't it showing a comparison with the rbm_softmax.py I just pushed to master?

File: rbm_softmax.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+        fe_corrupted = self._free_energy(v)
+        self.uncorrupt(v, state)
+        return fe.mean(), fe_corrupted.mean()
+
+        # TODO: I don't have a great intuition about this. Why multiply by n_features?
+        # The overfitting section of "Practical Guide" just says to compare the
+        # average free energy of train and validation and to compare them. Any reason
+        # not to just do that here as well? Seems much simpler to interpret. Maybe
+        # it's because we're supposed to be dealing with smaller deltas here?
+        #return v.shape[1] * log_logistic(fe_corrupted - fe)
```
Comment (colinmorris) [inline rbm_softmax.py:354]: @falsifian the commented line above is what was used in the original implementation. Any idea why it would be used rather than the seemingly-more-straightforward thing  I did? I found this number confusing to interpret ("if the energy of the training data goes down relative to the corrupted version, then the sigmoid goes to -1, so the log goes to 0 from the left, so...")

File: rbm_softmax.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+            self.intercept_visible_ = np.log( vis_priors / (1 - vis_priors) )
+
+
+        else:
+            print "Resuing existing weights and biases"
+        # Don't necessarily want to reuse h_samples if we have one leftover from before - batch size might have changed
+        self.h_samples_ = np.zeros((self.batch_size, self.n_components))
+
+        n_batches = int(np.ceil(float(n_samples) / self.batch_size))
+        batch_slices = list(gen_even_slices(n_batches * self.batch_size,
```
Comment (colinmorris) [inline rbm_softmax.py:413]: I just noticed batches aren't shuffled between epochs. But that doesn't really matter - right?

File: rbm_softmax.py [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+                begin = end
+
+        return self
+
+    def wellness_check(self, epoch, duration, train, validation):
+        """Log some diagnostic information on how the model is doing so far."""
+        validation_debug = ''
+        if validation is not None:
+            v_energy, t_energy = self.score_validation_data(train, validation)
+            # TODO: Not clear whether relative or absolute difference is the more relevant metric here.
```
Comment (colinmorris) [inline rbm_softmax.py]: @falsifian do you have an intuition about this? I prefer absolute difference, because subsequent epochs tend to increase that, while the relative difference sometimes stays the same or goes down. And I'm an optimist.

Comment (colinmorris) [inline rbm_softmax.py]: nvm, I think I figured it out.

p(v,h) = 1/Z \* e^{-E(v,h)}

So the ratio of the probabilities of v,h and v',h' is just e^{E(v,h) - E(v',h')}. So this is totally the right thing.

Comment (falsifian) [inline rbm_softmax.py]: Hey, my paper reviews are due Friday, so I'll definitely have time after
that! Maybe earlier

Comment (colinmorris) [inline rbm_softmax.py]: I mean, I was just saying I figured this one out. 

I'm actually going to go ahead and merge this today, because it's getting sillily big. But you're still welcome to take a look whenever.

File: common.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        line = line.strip()
+        try:
+            vecs.append(codec.encode(line))
+        except NonEncodableTextException as e:
+            # Too long, or illegal characters
+            skipped[e.reason] += 1
+
+    print "Gathered {} vectors. Skipped {} ({})".format(len(vecs), 
+        sum(skipped.values()), dict(skipped))
+    # TODO: Why default to dtype=float? Seems wasteful? Maybe it doesn't really matter. Actually, docs here seem inconsistent? Constructor docs say default float. transform docs say int.
```
Comment (falsifian) [inline common.py:43]: The [numpy.asarray docs](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.asarray.html) say default is inferred from the input. So I think it depends what the codec outputs.

File: common.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    # TODO: should probably try using a sparse matrix here
```
Comment (falsifian) [inline common.py:44]: I'm confused. Doesn't ShortTextCodec output dense vectors?

File: common.py (hunk 5 above)
Comment (falsifian) [inline common.py:43]: Oh, sorry, I thought this was applying to the `vecs = ...` line. Yeah, you should file a bug with [skikit-learn](https://github.com/scikit-learn/scikit-learn).

File: common.py (hunk 6 above)
Comment (falsifian) [inline common.py:44]: Oops, again, thought you were referring to the next line.

The OneHotEncoder constructor says `sparse = True` by default though.

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+*.pickle
+*.pyc
+*.html
+data/
```
Comment (falsifian) [inline .gitignore:4]: Probably shouldn't list `data/` if you have version-controlled files there.

You have a `.gitignore` in the `data` directory anyway.

File: data/len_histogram.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+# Print a histogram of line lengths for a text file
+awk '{ print length($0) }' $1 | sort | uniq -c | sort -n -r
```
Comment (falsifian) [inline data/len_histogram.sh:3]: It's better to write `"$1"` instead of `$1`.

Right now, if you run
  `len_histogram.sh 'file with spaces.txt'`
it will pass `file`, `with` and `spaces.txt` as separate arguments to awk.

Comment (falsifian) [inline data/len_histogram.sh:3]: BTW this is a legit reason to dislike writing shell scripts.

File: rbm_softmax.py (hunk 1 above)
Comment (falsifian) [inline rbm_softmax.py:1]: I think it's worthwhile to be able to see the diff. If you're as keen as I, I think the following three commits will do the trick:
1. delete `rbm_softmax.py`
2. add the one from sklearn
3. replace it with your version

File: rbm_softmax.py [hunk 9]
```diff
@@ -0,0 +1,529 @@
+"""Restricted Boltzmann Machine with softmax visible units.
+Based on sklearn's BernoulliRBM class.
+"""
+
+# Authors: Yann N. Dauphin <dauphiya@iro.umontreal.ca>
+#          Vlad Niculae
+#          Gabriel Synnaeve
+#          Lars Buitinck
+# License: BSD 3 clause
```
Comment (falsifian) [inline rbm_softmax.py:9]: Put a copy of the BSD 3 clause licence text somewhere.

File: rbm_softmax.py (hunk 1 above)
Comment (colinmorris) [inline rbm_softmax.py:1]: Good call. [Done in a branch.](https://github.com/colinmorris/char-rbm/commit/7d643993f40ee07f153048df0c879413260ee019)

(That's the very latest version of the file, so it's a bit messier than the one in this P.R.)

File: data/len_histogram.sh (hunk 8 above)
Comment (colinmorris) [inline data/len_histogram.sh:3]: As if I needed any more

File: sample.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    elif init_method == VisInit.uniform:
+        return np.random.randint(0, 2, vis_shape)
+    # This will fail if ' ' isn't in the alphabet of this model
+    elif init_method == VisInit.spaces or init_method == VisInit.padding:
+        fillchar = {VisInit.spaces: ' ', VisInit.padding: model.codec.filler}[init_method]
+        vis = np.zeros( (n,) + model.codec.shape())
+        try:
+            fill = model.codec.char_lookup[fillchar]
+        except KeyError:
+            raise BadInitMethodException(fillchar + " is not in model alphabet")
```
Comment (falsifian) [inline sample.py:51]: Nit: you probably want repr(fillchar) to avoid the message "  is not in model alphabet"
